README.md
# Flight Ticket Search Engine

## Overview
Welcome to the Flight Ticket Search Engine, an advanced tool created to assist users in finding customized and cost-effective travel itineraries. This application, built using the Clojure programming language, is designed to meet the varied requirements of our customers. It provides a smooth and effective method for locating the ideal flight schedule.

## Features

- **Personalized Search**: Enter precise parameters such as the city of departure, desired destination, and relevant passenger information to customize your trip arrangements.
- **Budget-Conscious Options**: Find travel arrangements that are in line with your financial limitations, guaranteeing cost-effectiveness.
- **Connection Constraints**: Discover flights that align with your desired number of connections to enhance your trip journey with greater ease and convenience.
- **Sorting by Total Price**: Search results are arranged in descending order based on the total price, giving priority to solutions that are more expensive but perhaps more convenient.
- **Interactive User Interface**: Experience a seamless search process with a UI that is both user-friendly and interactive.
- **Customer Feedback Loop**: We value your input! The search engine continuously evolves based on your feedback, ensuring a platform that meets your expectations.


## Technical Overview

The aircraft ticket search engine is constructed using a microservices architecture, making use of the functional programming features of Clojure. The search algorithm employs a Depth-First Search (DFS) methodology to uncover travel itineraries, taking into account budgetary and connectivity limitations for various categories of passengers.

## Getting Started

Follow these steps to get the flight ticket search engine up and running on your local machine:

### Clone the Repository

git clone [repository-url]
cd flight-ticket-search-engine


### Install Dependencies

# Assuming you have Leiningen installed

lein deps

### Run the Application

lein run


### Access the UI
Open your web browser and go to `src/flight/search_engine.clj` to access the user interface

## Contributing

We welcome contributions from the community! If you'd like to contribute to the flight ticket search engine, please follow our Contribution Guidelines.

## Issues and Feedback

At the moment, the program cannot gracefully handle invalid inputs (such as non-existent cities or negative budget values) or missing fields in the dataset.

## References

- Higginbotham, D. (2015). _Clojure for the Brave and True_. No Starch Press. [ISBN: 9781593275914](https://books.google.cz/books?id=mQLPCgAAQBAJ)
- Meier, C. (2015). _Living Clojure: An Introduction and Training Plan for Developers_. O’Reilly Media. [ISBN: 9781491909294](https://books.google.cz/books?id=b4odCAAAQBAJ)
- McDonnell, M. (2017). Quick Clojure. Effective Functional Programming. ISBN: 9781484229521. URL: https://rb.gy/wmvbfh
- Karumanchi, N. (2011). Data Structures and Algorithms Made Easy: Data Structure and Algorithmic Puzzles. CareerMonk Publications. ISBN: 9780615459813. URL: https://books.google.cz/books?id=FPIznwEACAAJ
- Rathore, A. (2015). Clojure in Action. ISBN: 9781638355335. URL: https://rb.gy/esew7a
- Naccache, R. (2015). Clojure Data Structures and Algorithms Cookbook. ISBN: 9781785287824. URL: https://rb.gy/eipm95


## License

This is academic work done under a free license.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2023-11-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2023-11-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://sourcehost.site/your-name/closure-project/compare/0.1.1...HEAD
[0.1.1]: https://sourcehost.site/your-name/closure-project/compare/0.1.0...0.1.1


test/clojure_project/core_test.clj
(ns closure-project.core-test
  (:require [clojure.test :refer :all]
            [closure-project.core1 :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/flight/search_engine.clj
(ns flight.search-engine
      (:require [clojure.string :as str]))

  ; Define a map of available flights where each key is a vector representing a route
  ; from a departure city to a destination city, and each value is a map with the price
  ; and number of connections for that flight.
  (def available-flights
       {["Krakov" "Warsaw"]     {:price 100 :connections 1}
        ["Hamburg" "Berlin"]    {:price 100 :connections 1}
        ["Warsaw" "Berlin"]     {:price 300 :connections 1}
        ["Prague" "Berlin"]     {:price 200 :connections 1}
        ["Munich" "Berlin"]     {:price 100 :connections 1}
        ["Munich" "Innsbruck"]  {:price 100 :connections 1}
        ["Vienna" "Innsbruck"]  {:price 200 :connections 1}
        ["Vienna" "Budapest"]   {:price 300 :connections 1}
        ["Warsaw" "Budapest"]   {:price 400 :connections 1}
        ["Zagreb" "Budapest"]   {:price 200 :connections 1}
        ["Vienna" "Rome"]       {:price 400 :connections 1}
        ["Napoli" "Rome"]       {:price 200 :connections 1}
        ["Napoli" "Rijeka"]     {:price 100 :connections 1}
        ["Vienna" "Prague"]     {:price 200 :connections 1}
        ["Vienna" "Rijeka"]     {:price 400 :connections 1}
        ["Rijeka" "Zagreb"]     {:price 100 :connections 1}
        ["Vienna" "Zagreb"]     {:price 300 :connections 1}
        ["Munich" "Zagreb"]     {:price 400 :connections 1}
        ["Innsbruck" "Rome"]    {:price 400 :connections 1}
        ["Budapest" "Rome"]     {:price 400 :connections 1}
        ["Budapest" "Berlin"]   {:price 300 :connections 1}
        ["Prague" "Brno"]       {:price 100 :connections 1}
        ["Prague" "Budapest"]   {:price 300 :connections 1}})

  (defn get-all-flight-hops [src dest]
        (letfn [(search [current-path visited?]
                  (let [current-city (peek current-path)]
                    (if (= current-city dest)
                      (list current-path)
                      (mapcat (fn [[[from to] _]]
                                (when (and (= current-city from)
                                           (not (visited? to)))
                                  (search (conj current-path to) (conj visited? to))))
                              available-flights))))]
          (search [src] #{src})))

  (defn get-all-flight-paths [src dest]
        (let [hops (get-all-flight-hops src dest)]
          (map (fn [hop]
                 (->> hop
                      (partition 2 1)
                      (map vec)
                      (select-keys available-flights)))
               hops)))

  ; Define the basic requirements for families and groups. Here, we use a map where the
  ; keys are the type of passenger group ('family' or 'group') and the values are maps
  ; that specify the maximum number of connections and the budget.
  (def requirements
       {:family {:max-connections 2 :max-budget 700}
        :group  {:max-connections 3 :max-budget 1000}})

  (defn route-price-and-connections [route]
        {:route-connections (->> route vals (map :connections) (reduce +))
         :route-price       (->> route vals (map :price) (reduce +))})

  ; Example function to check if a route meets the family or group requirements
  (defn route-meets-requirements?
        [route-type route]
        (let [{:keys [max-connections max-budget]} (requirements route-type)
              {:keys [route-connections route-price]} (route-price-and-connections route)]
          (and (<= route-connections max-connections)
               (<= route-price max-budget))))

  (defn find-flights
        "Returns a list of possible routes from `departure` to `destination` within the given `max-connections` and `max-price`."
        [departure destination type]
        (->> (get-all-flight-paths departure destination)
             (filter #(route-meets-requirements? type %))))

  (defn format-travel-plans
        "Formats a list of travel plans for display."
        [travel-routes]
    ; Define a local function to format a single travel plan
        (let [format-plan (fn [route]
                            (let [{:keys [route-connections route-price]} (route-price-and-connections route)]
                              (prn "--------" route)
                              (str "Travel Plan:\n"
                                   "  From: " (-> route first first first) "\n" ; Append the departure city
                                   "  To: " (-> route first first second) "\n" ; Append the destination city
                                   "  Total Price: $" route-price "\n" ; Append the total price
                                   "  Total Connections: " route-connections "\n")))] ; Append the total number of connections
          ; Apply the format-plan function to each travel plan in the list and concatenate the results into a single string
          (apply str (map format-plan travel-routes))))

  (defn parse-and-validate-input
        "Parses and validates user input. Expected input format: 'departure-city,destination-city,passenger-type'."
        [input-string]
        (let [[departure destination type] (clojure.string/split input-string #",")
              type (keyword type)]
          (println "[validate-input] departure =" departure "; destination =" destination "; type =" type)
          (cond
            (str/blank? departure)
            (do (println "Departure city is missing")
                false)

            (str/blank? destination)
            (do (println "Destination city is missing")
                false)

            (not (#{:family :group} type))
            (do
              (println "Group is missing")
              false)
            :else [departure destination type])))

  ;;-main function for launching from IDE also from command line
  (defn -main
        "The main function that ties everything together in the travel application."
        [& [input]]
        (let [input-string (or input "Prague,Berlin,family")]
          ; Parse and validate the input
          (when-let [[departure destination type] (parse-and-validate-input input-string)]
            (let [; Find feasible travel plans based on the criteria
                  travel-plans (find-flights departure destination type)
                  formatted-plan (format-travel-plans travel-plans)]
              (prn formatted-plan)
              formatted-plan))))

src/closure_project/Pet Adoption System.clj
(ns closure-project.Pet Adoption System)

(ns pet-adoption.core
  (:gen-class))

(def pet-db (atom [])) ;; Atom to store a vector of pets

(defn create-pet [name species age breed description available]
  {:name name
   :species species
   :age age
   :breed breed
   :description description
   :available available})

(defn list-available-pets []
  (filter :available @pet-db))

(defn adopt-pet [index]
  (let [pet (nth @pet-db index)]
    (if (= (:available pet) true)
      (do
        (swap! pet-db update-in [index] assoc :available false)
        (println "You have adopted" (:name pet))
        true)
      false)))

(defn interact-with-pet [index action user-info]
  (let [pet (nth @pet-db index)]
    (if (and (not= (:available pet) true) (:species pet))
      (let [action-result
            (case action
              :walk-dog (if (= (:species pet) "dog") true false)
              :cuddle-cat (if (= (:species pet) "cat") true false))
            pet-id (count @pet-db)]
        (if action-result
          (do
            (println "You have interacted with" (:name pet))
            (conj! pet-db
                   {:id (inc pet-id)
                    :user-info user-info
                    :pet-info pet
                    :action action})
            true)
          false))
      false)))

(defn main-menu []
  (println "Welcome to the Pet Adoption System")
  (println "1. List Available Pets")
  (println "2. Adopt a Pet")
  (println "3. Interact with a Pet")
  (println "4. Exit"))

(defn adopt-menu []
  (println "Adopt a Pet:")
  (let [available-pets (list-available-pets)
        num-pets (count available-pets)]
    (if (empty? available-pets)
      (println "No pets available for adoption.")
      (do
        (doseq [i (range num-pets)]
          (let [pet (nth available-pets i)]
            (println (str (inc i) ". " (:name pet) " - " (:species pet)))))
        (print "Enter the number of the pet you want to adopt: ")
        (let [choice (Integer. (read-line))]
          (if (and (<= choice num-pets) (> choice 0))
            (if (adopt-pet (dec choice))
              (println "Congratulations! You have adopted a pet.")
              (println "Sorry, the pet is no longer available for adoption."))
            (println "Invalid choice. Please select a valid number.")))))))

(defn interact-menu []
  (println "Interact with a Pet:")
  (print "Enter the pet's index: ")
  (let [index (Integer. (read-line))]
    (if (and (<= index (dec (count @pet-db)) (> index -1))
             (let [pet (nth @pet-db index)]
               (if (= (:species (:pet-info pet)) "dog")
                 (do
                   (print "Enter your information (First Name Last Name DateOfBirth Email Phone Address): ")
                   (let [user-info (read-line)]
                     (if (interact-with-pet index :walk-dog user-info)
                       (println "You have taken the dog for a walk.")
                       (println "Sorry, you can't walk this pet."))
                     (print "Press any key to continue...")
                     (read-line)))
                 (if (= (:species (:pet-info pet)) "cat")
                   (do
                     (print "Enter your information (First Name Last Name DateOfBirth Email Phone Address): ")
                     (let [user-info (read-line)]
                       (if (interact-with-pet index :cuddle-cat user-info)
                         (println "You have spent time with the cat.")
                         (println "Sorry, you can't spend time with this pet."))
                       (print "Press any key to continue...")
                       (read-line))))))))
    (do
      (println "Invalid index. Please enter a valid index.")
      (print "Press any key to continue...")
      (read-line)))))

(defn -main [& args]
  (loop []
    (main-menu)
    (print "Enter your choice (1/2/3/4): ")
    (let [choice (Integer. (read-line))]
      (cond
        (= choice 1) (adopt-menu)
        (= choice 2) (adopt-menu)
        (= choice 3) (interact-menu)
        (= choice 4) (System/exit 0)
        :else (println "Invalid choice. Please enter a valid choice."))
      (recur))))

(-main)


src/ica/ica1.clj
(ns ica.ica1
  (:require [clojure.string :as str]))
; Define a map of available flights where each key is a vector representing a route
; from a departure city to a destination city, and each value is a map with the price
; and number of connections for that flight.
(def flights
  {["Prague" "Vienna"] {:price 100 :connections 1}
   ["Vienna" "Zadar"]  {:price 200 :connections 1}
   })

; Define the basic requirements for families and groups. Here, we use a map where the
; keys are the type of passenger group ('family' or 'group') and the values are maps
; that specify the maximum number of connections and the budget.
(def requirements
  {:family {:max-connections 2 :max-budget 700}
   :group  {:max-connections 3 :max-budget 1000}})

; Example function to check if a route meets the family or group requirements
(defn route-meets-requirements?
  [route-type route]
  (let [reqs (requirements route-type)]
    (and (<= (:connections route) (:max-connections reqs))
         (<= (:price route) (:max-budget reqs)))))

(defn find-flights
  "Returns a list of possible routes from `start` to `end` within the given
  `max-connections` and `max-price`."
  [start end max-connections max-price]
  ; A helper function to extend the current path with a new flight.
  (letfn [(extend-path [current-path flight]
            (let [current-city (first flight)
                  next-city (second flight)
                  flight-info (flights flight)]
              (conj current-path
                    {:from current-city
                     :to next-city
                     :price (:price flight-info)
                     :connections (:connections flight-info)})))]
    ; The core recursive DFS function.
    (loop [paths-to-visit (list [{:path [] :total-price 0 :total-connections 0}])
           valid-paths []]
      (if (empty? paths-to-visit)
               ; If there are no more paths to visit, return the valid paths found
               valid-paths
               ; Otherwise, continue with DFS
               (let [{:keys [path total-price total-connections]} (first paths-to-visit)
                     remaining-paths (rest paths-to-visit)
                     current-city (if (empty? path) start (:to (last path)))]
                 (if (= current-city end)
                   ; If the current path has reached the end, add it to the valid paths
                   (recur remaining-paths (conj valid-paths path))
                   ; If not at the end, find all possible next flights and extend the path
                   (let [; Find flights departing from the current city.
                         next-flights (filter #(= (first %) current-city) (keys flights))

                         ; Filter out flights leading to already visited cities in the path
                         new-destinations (filter #(not (contains? (set (map :to path)) (second %))) next-flights)

                         ; Further filter out any flights that exceed budget or connection constraints
                         feasible-flights (filter #(let [info (flights %)]
                                                     (and (<= (+ total-price (:price info)) max-price)
                                                          (<= (+ total-connections (:connections info)) max-connections)))
                                                  new-destinations)]

                   (recur (concat (map #(assoc % :path (extend-path path %)
                                                   :total-price (+ total-price (:price (flights %)))
                                                   :total-connections (+ total-connections (:connections (flights %))))
                                         feasible-flights)
                                    remaining-paths)
                            valid-paths))))))))

(defn parse-and-validate-input
  "Parses and validates user input. Expected input format: 'departure-city,destination-city,passenger-type'."
  [input-string]
  (let [parsed-input (clojure.string/split input-string #",")
        [departure destination type] parsed-input]

    ; Validate the input components
    (when (or (nil? departure) (empty? departure))
      (throw (IllegalArgumentException. "Departure city is missing")))

    (when (or (nil? destination) (empty? destination))
      (throw (IllegalArgumentException. "Destination city is missing")))

    (when (not (contains? #{:family :group} (keyword type)))
      (throw (IllegalArgumentException. "Passenger type must be either 'family' or 'group'")))

    ; Return a map of the parsed and validated input
    {:departure departure :destination destination :type (keyword (clojure.string/lower-case type))}

    (defn format-travel-plans
  "Formats a list of travel plans for display."
  [travel-plans]
  ; Define a local function to format a single travel plan
  (let [format-plan (fn [plan]
                      (str "Travel Plan:\n"
                           "  From: " (:departure plan) "\n"  ; Append the departure city
                           "  To: " (:destination plan) "\n"  ; Append the destination city
                           "  Total Price: $" (:total-price plan) "\n"  ; Append the total price
                           "  Total Connections: " (:total-connections plan) "\n"))]  ; Append the total number of connections
    ; Apply the format-plan function to each travel plan in the list and concatenate the results into a single string
    (apply str (map format-plan travel-plans))))

; Assume the following functions are defined:
; - parse-and-validate-input: Parses and validates user input
; - find-flights: Finds feasible flights based on criteria
; - format-travel-plans: Formats the list of travel plans for display

(defn main-function
  "The main function that ties everything together in the travel application."
  [input-string]
  (try
    ; Parse and validate the input
    (let [input-map (parse-and-validate-input input-string)
          departure (:departure input-map)
          destination (:destination input-map)
          type (keyword (:type input-map))
          ; Retrieve the requirements based on the type of passengers (family or group)
          reqs (requirements type)
          max-connections (:max-connections reqs)
          max-price (:max-budget reqs)
          ; Find feasible travel plans based on the criteria
          travel-plans (find-flights departure destination max-connections max-price)]

      ; Format and return the travel plans for display
      (format-travel-plans travel-plans))

    ; Catch and handle IllegalArgumentExceptions
    (catch IllegalArgumentException e
      (str "Error: " (.getMessage e)))))

; First, require our namespace
; (require 'ica.ica1)

; (require '[ica.ica1 :refer :all])


; Now, you can call the main function with different inputs to test it
; (ica.ica1/main-function "Prague,Zadar,family"